src/uu/mkdir/src/mkdir.rs

```diff
@@ -231,15 +232,52 @@ fn create_dir(path: &Path, is_parent: bool, config: &Config) -> UResult<()> {
         return Ok(());
     }
 
+    // Iterative implementation: collect all directories to create, then create them
+    // This avoids stack overflow with deeply nested directories
     if config.recursive {
-        match path.parent() {
-            Some(p) => create_dir(p, true, config)?,
-            None => {
-                USimpleError::new(1, translate!("mkdir-error-failed-to-create-tree"));
+        // Collect all parent directories that need to be created
+        let mut dirs_to_create = Vec::new();
+        let mut current = path;
+
+        // Walk up the tree collecting non-existent directories
+        while let Some(parent) = current.parent() {
+            if parent == Path::new("") {
+                break;
             }
+            // Only check exists() for paths without ".." to avoid false positives
+            // ("test_dir/.." may return true even when the logical parent doesn't exist)
+            // Use as_os_str() to avoid string allocation which can use stack space
+            let has_dotdot = parent
+                .as_os_str()
+                .as_encoded_bytes()
+                .windows(2)
+                .any(|w| w == b"..");
```

Comment:
It might be clearer to use `components()`:
```suggestion
            let has_dotdot = parent.components().any(|c| c.as_os_str() == "..");
```